src/prototype_room_utils.js

```diff
@@ -133,24 +133,67 @@ Room.prototype.checkRoleToSpawn = function(role, amount, targetId, targetRoom) {
   this.memory.queue.push(creepMemory);
 };
 
-Room.prototype.getPartConfig = function(energy, parts) {
-  var sum = 0;
-  var i = 0;
-  var partConfig = [];
-  while (sum < energy && partConfig.length < 50) {
-    var part = parts[i % parts.length];
-    if (sum + BODYPART_COST[part] <= energy) {
-      partConfig.push(part);
-      sum += BODYPART_COST[part];
-      i += 1;
-    } else {
-      break;
-    }
-  }
-  return partConfig;
+Room.prototype.getPartConfig = function(datas) {
+
+  let layout = datas.layout;
+  let amount = datas.amount;
+  let minEnergyStored = datas.minEnergyStored;
+  let maxEnergyUsed = datas.maxEnergyUsed;
+  let bonusParts = datas.bonusParts;
```

Comment:
bonusParts is actually a "prefix" but actually, on final body, all is sorted by bodytype. don't know if i'll not find something better. i'll think about builds because it's true that acutally, there are some roles with the sames layout.